components/engine/src/rocks/util/io_limiter.rs

```diff
@@ -105,14 +105,54 @@ impl<'a, T: Write + 'a> Write for LimitWriter<'a, T> {
     }
 }
 
+/// A limited reader.
+///
+/// The read limits the bytes per second read from an underlying reader.
+pub struct LimitReader<'a, T: Read> {
+    limiter: Option<Arc<IOLimiter>>,
+    reader: &'a mut T,
+}
+
+impl<'a, T: Read + 'a> LimitReader<'a, T> {
+    /// Create a new `LimitReader`.
+    pub fn new(limiter: Option<Arc<IOLimiter>>, reader: &'a mut T) -> LimitReader<'a, T> {
+        LimitReader { limiter, reader }
+    }
+}
+
+impl<'a, T: Read + 'a> Read for LimitReader<'a, T> {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        if let Some(ref limiter) = self.limiter {
+            let total = buf.len();
+            let single = limiter.get_max_bytes_per_time() as usize;
+            let mut count = 0;
+            let mut curr = 0;
+            let mut end;
+            while curr < total {
+                if curr + single >= total {
+                    end = total;
+                } else {
+                    end = curr + single;
+                }
+                limiter.request((end - curr) as i64);
```

Comment:
why not call `request` with `total`? if it doesn't satisfy, the call would be blocked.

Reply:
It aborts if `total > limiter.get_max_bytes_per_time()`.